### ISSUE TITLE
How advanced skinning are implemented 

### ISSUE THREAD

**Issue by CppProgrammer23** [KhronosGroup/glTF-Sample-Models#312]
Hi,

I want to implement an algorithm for advanced skinnig. So I have few questions; first, how can i know which animations start first (suppose that the rotation starts then the translation comes but both transformations have the same key frames and the translation is 0 and only changes after 2 seconds.)?
If i understand, I have to access each animation and change the mode that it belongs to, accordingly to the value at that key frame?

**Comment by javagl** [comment]
It sounds like this is only implicitly about skinning, but primarily about *animation*. Specifically, the question

> how can i know which animations start first

sounds like what is "explained" in the Notes at https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#animations : There may be multiple animations, and they are supposed to be usable independently. This may also answer the question about the "duration" of animations. But note that there is a caveat regarding the *channels* of **one** animation. This is explained at https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/BoxAnimated (and this is a good test model to start with basic animations). Essentially, when there are multiple channels with different lengths, then the longest one determines the length of the animation, and the individal channels may not "wrap around" or so. 

In order to test basic skinning+animation, https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/RiggedSimple may be good start.

**Comment by CppProgrammer23** [comment]
sounds great. I still have a confusion with this formula: 
`jointMatrix(j) =`
  `globalTransformOfNodeThatTheMeshIsAttachedTo^-1 *`
  `globalTransformOfJointNode(j) *`
 ` inverseBindMatrixForJoint(j);`

the 2 last variables   it's ok, I understood them, but I struggle with the first one.
imagine I have this configuration:
`"nodes" : [ {`
    `"skin" : 0,`
    `"mesh" : 0`
 ` }, {`
    `"children" : [ 2 ],`
    `"translation" : [ 0.0, 1.0, 0.0 ]`
  `}, {`
    `"rotation" : [ 0.0, 0.0, 0.0, 1.0 ]`
 ` } ]`

So the `globalTransformOfNodeThatTheMeshIsAttachedTo^-1` would be the inverse of `translation = 0`, the `rotation = default (no rotation)` and the `scale = 1` because this node doesn't implement any transformations. but it doesn't work as that way. I get something wrong. When I tried  `translation = 0, 1, 0` as in the `first node`, `rotation as in node 1` and `scale = 1` it does give me something right.

Imagine I have this model:

`"nodes": [`
  `{`
      `"name": "Light",`
      `"rotation": [`
        `0.16907575726509094,`
        `0.7558803558349609,`
        `-0.27217137813568115,`
       `0.570947527885437`
      `],`
      `"translation": [`
        `4.076245307922363,`
        `5.903861999511719,`
        `-1.0054539442062378`
      `]`
   ` },`
    `{`
      `"name": "Camera",`
      `"rotation": [`
        `0.483536034822464,`
        `0.33687159419059753,`
        `-0.20870360732078552,`
       `0.7804827094078064`
      `],`
      `"translation": [`
       `7.358891487121582,`
        `4.958309173583984,`
       `6.925790786743164`
     `]`
    `},`
    `{`
      `"name": "Bone.00",`
      `"rotation": [`
        `-0.0020970599725842476,`
        `-0.001136002130806446,`
        `1.6966068869805895e-05,`
        `0.9999971389770508`
     `],`
      `"scale": [`
       `0.9999999403953552,`
       `0.9999997615814209,`
        `0.9999998807907104`
      `],`
      `"translation": [`
        `3.7182687484538235e-12,`
        `1.3126283884048462,`
        `-2.1039745501383322e-09`
      `]`
    `},`
    `{`
      `"children": [`
        `2`
     `],`
      `"name": "Bone.01",`
      `"rotation": [`
        `0.011737177148461342,`
        `0.002114036586135626,`
        `-0.00010634021600708365,`
        `0.999928891658783`
     ` ],`
      `"scale": [`
        `1,`
        `1,`
        `0.9999999403953552`
      `],`
      `"translation": [`
        `3.3658409392955946e-11,`
        `0.8158659338951111,`
        `1.369617641522325e-09`
      `]`
    `},`
    `{`
      `"children": [`
        `3
      `],`
      `"name": "Bone.02",`
      `"rotation": [`
        `-0.5002185702323914,`
        `0.49416103959083557,`
        `0.505838930606842,`
        `0.4997132420539856`
      `],`
      `"scale": [`
        `1,`
        `1,`
        `1.0000001192092896`
      `],`
      `"translation": [`
        `3.1174166202545166,`
        `-0.02734565921127796,`
        `-3.7103995431664316e-09`
      `]`
    `},`
    `{`
      `"mesh": 0,`
      `"name": "Plane",`
      `"skin": 0`
    `},`
    `{`
      `"children": [`
        `4`
      `],`
      `"name": "Armature",`
      `"rotation": [`
        `-0.4999999701976776,`
        `-0.5,`
        `0.5,`
        `-0.5000000596046448`
      `],`
      `"translation": [`
        `-0.866496205329895,`
        `-1.5603487491607666,`
        `1.647308349609375`
      `]`
   `}`
  `]`

This is an example that I've created under Blender. So I wonder how to get the `globalTransformOfNodeThatTheMeshIsAttachedTo^-1`? is it from the node 5 (starting from 0) and how to propagate then to get it?

**Comment by CppProgrammer23** [comment]
thanks for this tip. I though it's only available on Stack Overflow 😄 

I attached the model with its bin file.
[skinTube.zip](https://github.com/KhronosGroup/glTF-Sample-Models/files/6460413/skinTube.zip)

I really want to implement the algorithm to make this model working: https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/CesiumMan

the model that you pointed in the previous comment (Rigged Simple) uses `matrix` and my code didn't support this feature yet, but I'm planning to implement it once I finish with `T, R, S` because `matrix` needs one more job that I have to decompose it into `T, R, S`.

Now I want to describe you what I understood from skinning:
first, I have to get the global inverse transform which is a constant and this is from the node that is attached to the mesh (in other terms, the root node). naturally, I did implement this method.
Next, I have to update the global transform each time I have an animation done. But here, I have to take care about the children. If a parent has a child that's mean that the transformation of that parent has to be transmitted to the child? so when the child has a child, I have to propagate that logic until no child is present? here I have the idea to use a linked list.
 here you can find my skinned animation class: https://github.com/CppProgrammer23/skin-animation . can you please help me to fix it (there are no errors but I want to fix the logic)

**Comment by scurest** [comment]
You don't need to multiply by `globalTransformOfNodeThatTheMeshIsAttachedTo^-1`. It's only purpose is to cancel out the multiplication by `globalTransformOfNodeThatTheMeshIsAttachedTo` that is applied to the vertices of unskinned meshes. For skinned meshes, just don't multiply by `globalTransformOfNodeThatTheMeshIsAttachedTo`, and you won't need to multiply by (or even calculate) the `globalTransformOfNodeThatTheMeshIsAttachedTo^-1`.

**Comment by CppProgrammer23** [comment]
Hi,

I tried to implement something close to what you said but I still cannot get what I want. I'm still not understanding how skins work.
I've updated my repository https://github.com/CppProgrammer23/skin-animation with new code (I cleaned the unused piece of code). I hope you can help me (by pulling some code to make some change or giving some piece of code). One month now struggling with Skins and still not understanding how it works. I've already implemented in my loader the camera, the animations (without skinning), the textures, .. it only misses the skins attribute. can you please help me _**Marco**_ ?

**Comment by javagl** [comment]
> For skinned meshes, just don't multiply by globalTransformOfNodeThatTheMeshIsAttachedTo

It might be possible to somehow sneak around that for the skinning computations themsefl (I'd have to scribble a down the math with pencil+paper to be "more sure" about that). But at *some* point, the `globalTransformOfNodeThatTheMeshIsAttachedTo` **will** be taken into account - otherwise, you couldn't move the skinned object around in the scene. So... 1. are you *sure* that this is possible? 2. did you implement it like that? and 3. can you still move around your skinned models by applying a transform, e.g. to the root node?

(An aside: I haven't checked that, but it might well be that **none** of the skinning test models uses a global transform that is **not** the identity matrix. If this is the case, it could give implementors the impression that their skinning works, even though their implementation might fail when the model is transformed. This has to be analyzed further, and *if* this is the case, it might be worth adding a dedicated sample model to test this...)

**Comment by javagl** [comment]
@CppProgrammer23 

Don't dispair. The skinning is the hardest part (or... well... at least, before PBR was introduced). I had a short look at the code, but again: It's hard to figure out the intention of some parts, or why they are implemented like that. For example, a function like the `glm::mat4 SkinnedAnimation::getGlobalInverseTransform(tinygltf::Scene scene)` looks a bit odd: It determines *some* node as the `rootNode`, and from whichly looking over it, it computes "something", but certainly not a global transform. The `traverseNode` function is not called, from what I can see. 

---

And by the way: It may be a matter of style, but I don't see a reason why someone should implement a function like

    Matrix getGlobalInverseTransform(Node node) {
        // many lines of code
        ...
        return glm::inverse(resultMatrix);
    }

instead of

    Matrix getGlobalTransform(Node node) {
        // many lines of code
        ...
        return resultMatrix;
    }
    Matrix getInverseGlobalTransform(Node node) {
        return glm::inverse(getGlobalTransform(node));
    }

I mean, the function to compute the global transform of a node will be required anyhow. And... of course, this function should not check for a certain `node.name`, or modify the state of the object that it is a member of, ... but... all I want to say is: It's hard to understand the goals of the current code, for me...

**Comment by CppProgrammer23** [comment]
When I wrote some _specific_ code for the model that you have created `(simpleSkin)`, it worked for me, but now I added this class `SkinnedAnimation` to make it generic for many models. I did understand how it worked when you explained it to me in the previous issue. Naturally, I have understand many parts in the Skinned animation but still confused with some parts. for example: how can I calculate the `globalInverseTransform` to be generic for every model? As it's mention in the tutorial, it's the inverse of the global Transform of node that the mesh is attached to, Imagine I have many roots in my glTF model as the model I linked in the previous comment (Camera, Light, etc..). How can I get the real root? this is exactly what make me confused. 
Naturally, I updated my code here: https://github.com/CppProgrammer23/skin-animation to call ``traverseNode` in `startAnimation` (when animation starts). can you just take a look on those methods and tell me what do you think. The methods are: `traverseNode` and `startAnimation`. The rest are just `Getters` and/or `Setters`.

What I really can't understand is in your tutorial about `Skinning` you mention that the `JointMatrix` is variable in function of the `globalTransform` and the `inverseBindMatrices`. For me, the `globalInverseTransform` is constant. And if you can describe me how can I get the `globalInverseTransform` (to be generic)

**Comment by scurest** [comment]
@javagl We've discussed this [before](https://github.com/KhronosGroup/glTF/issues/1784#issuecomment-606096424). Yes I am sure. About skinning the spec says "...while ignoring the transform of the skinned mesh node". You also [wrote](https://github.com/KhronosGroup/glTF-Tutorials/blame/87453aaaae91f47ce8146981b67cb11a2f8a4d1a/gltfTutorial/gltfTutorial_020_Skins.md#L126) that it's only purpose was to cancel out part of the modelview matrix. Yes, the root node still affects the skinned mesh, because it affects `globalTransformOfJointNode`.

**Comment by javagl** [comment]
@scurest You're right, but I still did not do all the maths behind that. I did that back when I wrote the tutorial and the overview, being completely new to vertex skinning, and digging through online resources and the COLLADA specification. But there had been some discussions, updates, clarifications and (apparently) simplifications in the meantime that I had not yet taken into account.

But after doing some archeology ( https://github.com/KhronosGroup/glTF/issues/1270 , https://github.com/KhronosGroup/glTF/issues/1403 , and more recently, the one that you now linked to), I *think* that you're right. I have read a bit in these issues, and tried out different cases (e.g. https://bghgary.github.io/glTF-Assets-Viewer/?manifest=https://raw.githubusercontent.com/KhronosGroup/glTF-Asset-Generator/v0.6.0/Output/Manifest.json&folder=2&model=10 ), with my implementation, once with and once without the global transform being integrated, and the result seems to be the same.

(I'm still not "scientist-sure" about that, because I didn't do the math all the way down, but ... I'm "engineer-sure": It seems to be correct from all what I've read and tried out...)

This means that the tutorial, the overview (and my implementation) have to be updated accordingly. I'll open an issue for that (but cannot say for sure when I'll be able to tackle that - I've been doing far too little for my "real" work recently anyhow...).
